File: SwiftApp/SwiftApp/Track.swift
//
//  Copyright (c) 2015 Harry Cheung
//

import UIKit

final class Track {
  
  var gates: [Gate] = []
  let id: Int!
  let name: String!
  var start: Gate!
  
  init(jsonObject: NSDictionary) {
    let jsonTrack = jsonObject["track"] as! NSDictionary
    let jsonGates = jsonTrack["gates"] as! NSArray
    for i in 0..<jsonGates.count {
      let jsonGate = jsonGates[i] as! NSDictionary
      let gate = Gate(
        type:        GateType(rawValue: jsonGate["type"] as! String)!,
        splitNumber: jsonGate["split_number"]!.integerValue,
        latitude:    jsonGate["latitude"]!.doubleValue,
        longitude:   jsonGate["longitude"]!.doubleValue,
        bearing:     jsonGate["bearing"]!.doubleValue)
      if gate.type == GateType.START_FINISH ||
        gate.type == GateType.START {
          start = gate
      }
      gates.append(gate)
    }
    id   = jsonTrack["id"]?.integerValue
    name = jsonTrack["name"] as! String
    
    assert(id != nil)
    assert(name != nil)
    assert(start != nil)
  }
  
  convenience init(jsonData: NSData) {
    let jsonObject = NSJSONSerialization.JSONObjectWithData(jsonData, options: NSJSONReadingOptions.MutableContainers, error: nil) as! NSDictionary
    self.init(jsonObject: jsonObject)
  }
  
  convenience init(file filePath: String) {
    var error:NSError?
    var content = NSData(contentsOfFile: filePath)
    // TODO: do something with error
    self.init(jsonData: content!)
  }
  
  func numSplits() -> Int {
    return gates.count
  }
  
  func distanceToStart(#latitude: Double, longitude: Double) -> Double {
    return start.location.distanceTo(Point(latitude: latitude, longitude: longitude));
  }
}

File: Titanium/Resources/track.js
'use strict';

//
//  Copyright (c) 2015 Harry Cheung
//

if (typeof require !== 'undefined') {
	var Gate = require('gate');
}

function Track(json) {
	this.start = null;
	this.gates = [];
	var jsonTrack = json.track;
	var jsonGates = jsonTrack.gates;
	var length = jsonGates.length;
	for (var i = 0; i < length; i++) {
		var jsonGate = jsonGates[i];
		var gate = new Gate(jsonGate.gate_type,
			parseInt(jsonGate.split_number),
			parseFloat(jsonGate.latitude),
			parseFloat(jsonGate.longitude),
			parseFloat(jsonGate.bearing));
		if (gate.type === 'START_FINISH' || gate.type === 'START') {
			this.start = gate;
		}
		this.gates.push(gate);
	}
	this.id = parseInt(jsonTrack.id);
	this.name = jsonTrack.name;
}

Track.prototype.numSplits = function () {
	return this.gates.length;
};

if (typeof exports !== 'undefined' && typeof module !== 'undefined' && module.exports) {
	exports = module.exports = Track;
}
